### PR TITLE
perf: lazy initialise Diffie-Hellman in handshake

### PR DESCRIPTION
**Pull request by tearfur**
Avoid generating a Diffie-Hellman private key when we are using a plain connection.

### CODE REVIEW

**Review by Coeur — APPROVED** [review]

**Comment by reardonia** [comment]
I'm not at all sophisticated in C++, but why not use std::optional idiom instead of the introducing a new is_dummy idiom?

**Comment by tearfur** [comment]
I agree that `std::optional` would be better here.

`is_dummy()` was introduced to workaround a bug in `small::vector` that requires the `value_type` to be `DefaultInsertable`. See https://github.com/alandefreitas/small/pull/46.

Now that I think about it again, I'd rather use `std::vector` as a slower workaround before it gets fixed. c2a3bf2

File: tests/libtransmission/crypto-test.cc [hunk 1]
```diff
@@ -48,25 +48,25 @@ std::string toString(std::array<std::byte, N> const& array)
 
 TEST(Crypto, DH)
 {
-    auto a = tr_message_stream_encryption::DH{};
-    auto b = tr_message_stream_encryption::DH{};
+    auto a = tr_message_stream_encryption::DH{ tr_message_stream_encryption::DH::randomPrivateKey() };
```
Comment (nevack) [inline tests/libtransmission/crypto-test.cc:51]: Is `tr_message_stream_encryption::DH::randomPrivateKey()` used anywhere outside this test now?

**Review by nevack — APPROVED** [review]
LGTM

File: tests/libtransmission/crypto-test.cc (hunk 1 above)
Comment (tearfur) [reply tests/libtransmission/crypto-test.cc:51]: Yes.

https://github.com/transmission/transmission/blob/235736dcefa697bba4c1cc1ece119962bbb5950e/libtransmission/handshake.h#L77-L80

**Review by reardonia — CHANGES_REQUESTED** [review]
since you are already making some changes in handshake.h, maybe implement full tracking of the different Pads (A/B/C/D) ? send/recv is context specific, so use same pad_x_len_ naming and let context determine which is sent/recv.

requires either returning or otherwise setting the sent pad length in `send_public_key_and_pad`. 

I did this locally and found it helpful in debugging to see the sent-pad-len in logs.

**Review by reardonia — APPROVED** [review]
tested, running well.  with #6950 fixes important longstanding bug.

**Review by ckerr — APPROVED** [review]

**Pull request merged**